cpu/stm32/periph/adc_f3.c

```diff
@@ -0,0 +1,165 @@
+ * @author      Hugues Larrive <hugues.larrive@laas.fr>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/adc.h"
+#include "periph_conf.h"
+#include "xtimer.h"
+
+#define SMP_SLOW    (0x2) /*< Sampling time for slow channels
+                                (0x2 = 4.5 ADC clock cycles) */
+
+/**
+ * @brief   Allocate locks for all available ADC devices
+ */
+static mutex_t locks[ADC_DEVS];
+
+static inline ADC_TypeDef *dev(adc_t line)
+{
+    return (ADC_TypeDef *)(ADC1_BASE + (adc_config[line].dev << 8));
+}
+
+static inline void prep(adc_t line)
+{
+    mutex_lock(&locks[adc_config[line].dev]);
+/* Enable the clock here only if it will be disabled by done, else just
+ * enable it once in adc_init() */
+#if defined(RCC_AHBENR_ADC1EN)
+    periph_clk_en(AHB, (RCC_AHBENR_ADC1EN << adc_config[line].dev));
+#endif
+}
+
+static inline void done(adc_t line)
+{
+/* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
+ * don't disable the clock as the other device may still use it. */
+#if defined(RCC_AHBENR_ADC1EN)
+    periph_clk_dis(AHB, (RCC_AHBENR_ADC1EN << adc_config[line].dev));
+#endif
+    mutex_unlock(&locks[adc_config[line].dev]);
+}
+
+int adc_init(adc_t line)
+{
+    /* Check if the line is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
+    }
+
+    /* Lock device and enable its peripheral clock */
+    prep(line);
+/* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
+ * enable the clock only once here. */
+#if !defined(RCC_AHBENR_ADC1EN)
+    periph_clk_en(AHB, (RCC_AHBENR_ADC12EN << adc_config[line].dev));
+#endif
+
+    /* Setting ADC clock to HCLK/1 is only allowed if AHB clock
+     * prescaler is 1 */
+    if (!(RCC->CFGR & RCC_CFGR_HPRE_3)) {
+        /* set ADC clock to HCLK/1 */
+        ADC12_COMMON->CCR |= ADC_CCR_CKMODE_0;
+    }
+    else {
+        /* set ADC clock to HCLK/2 otherwise */
+        ADC12_COMMON->CCR |= ADC_CCR_CKMODE_1;
+    }
+
+    /* Configure the pin */
+    gpio_init_analog(adc_config[line].pin);
+
+    /* Init ADC line only if it wasn't already initialized */
+    if (!(dev(line)->CR & ADC_CR_ADEN)) {
+        /* Enable ADC internal voltage regulator and wait for startup period */
+        dev(line)->CR |= ADC_CR_ADVREGEN;
+        xtimer_usleep(ADC_T_ADCVREG_STUP_US);
+
+        /* Configure calibration for single ended input */
+        dev(line)->CR &= ~ADC_CR_ADCALDIF;
+
+        /* Start automatic calibration and wait for it to complete */
+        dev(line)->CR |= ADC_CR_ADCAL;
+        while (dev(line)->CR  & ADC_CR_ADCAL) {}
+
+        /* Clear ADRDY by writing it */
+        dev(line)->ISR |= ADC_ISR_ADRDY;
+
+        /* Enable ADC and wait for it to be ready */
+        dev(line)->CR |= ADC_CR_ADEN;
+        while ((dev(line)->ISR & ADC_ISR_ADRDY) == 0) {}
+
+        /* Set sequence length to 1 conversion */
+        dev(line)->SQR1 |= (0 & ADC_SQR1_L);
+    }
+
+    /* Configure sampling time for the given channel (6 to 18) */
```

Comment:
We should add that to the ADC configuration in the future - let's keep that in mind for `adc_ng`.